### PR TITLE
Port message export API

### PR DESCRIPTION
**Pull request by JayH5**
See issue #5. This is the first HTTP API for the new message store

Not ready to merge - a test fails
- Mostly copied over from main Vumi project
- "concurrency" option removed
- Small cleanup/refactor in some places
- Some more tests added for bad requests (not all passing)

### CODE REVIEW

File: vumi_message_store/api/message_export_resources.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        Process a page of keys and each subsequent page.
+
+        The keys for the current page are handed off to :meth:`fetch_page` for
+        processing. If there is another page, we fetch that while the current
+        page is being handled and add a callback to process it when the
+        current page is finished.
+
+        When there are no more pages, we add a callback to close the request.
+        """
+        if request.connection_has_been_closed:
```
Comment (JayH5) [inline vumi_message_store/api/message_export_resources.py:99]: This branch is currently the only hole in the test coverage

File: vumi_message_store/api/tests/test_message_export_worker.py [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+
+        self.assertEqual(resp.code, http.NOT_FOUND)
+
+    @inlineCallbacks
+    def test_get_invalid_batch(self):
+        """
+        An HTTP request to the server for a non-existant batch should return a
+        404 response code.
+        """
+        # TODO: This test fails. Returns 200 + blank response
```
Comment (JayH5) [inline vumi_message_store/api/tests/test_message_export_worker.py]: Is there any way to determine if a batch exists before we do a `list_batch...` operation? Because what happens is that we do a `list_batch`, there are no pages to iterate so we just return... and return a 200 response code (as far as I can see).

File: vumi_message_store/api/message_export_resources.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+from vumi_message_store.api.message_export_formatters import (
+    JsonFormatter, CsvFormatter)
+from vumi.message import format_vumi_date
+
+
+class ParameterError(Exception):
+    """
+    Exception raised while trying to parse a parameter.
+    """
+    pass
```
Comment (jerith) [inline vumi_message_store/api/message_export_resources.py]: You don't need `pass` if you have a docstring.

File: vumi_message_store/api/tests/test_message_export_worker.py (hunk 2 above)
Comment (jerith) [inline vumi_message_store/api/tests/test_message_export_worker.py]: There is a way, but it requires a Redis call. We can assume that anyone calling this API already has something that they think is a valid batch identifier and for message listing purposes there isn't much difference between an empty batch and a non-existent one.

**Comment by jerith** [comment]
Looks good. Testing the connection lost case is tricky, so let's maybe add a skip marker to that test and handle it in a new branch.

**Comment by JayH5** [comment]
@jerith: how do I add a skip marker?

**Comment by jerith** [comment]
Add a `skip` attribute to the method. https://github.com/praekelt/vumi/blob/develop/vumi/persist/tests/test_model.py#L501-L502 is an example of where we do this.

**Comment by jerith** [comment]
Builds are failing because `MessageExportResourceWorker` isn't properly cleaning up its Riak and Redis managers. (The check for this is currently optional, but it's going to become mandatory at some point.) You should run tests with the environment variable `VUMI_TEST_ASSERT_CLOSED=true` set.

**Comment by jerith** [comment]
I've taken the liberty of pushing a commit to fix the manager cleanup. I'm happy with the rest of the branch, although you may want to rename `MessageExportResourceWorker` to something a little less cumbersome. (Perhaps drop the `Resource` or replace it with `Service` or something?)

If you're happy with my commit you can merge this branch as-is or do the rename and ping me for a last quick review.

**Comment by JayH5** [comment]
@jerith thanks! I agree.. it is a bit wordy

File: vumi_message_store/api/tests/test_message_export_worker.py [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+        yield worker.startService()
+        port = yield worker.services[0]._waitingForPort
+        addr = port.getHost()
+
+        self.url = 'http://%s:%s' % (addr.host, addr.port)
+
+        self.query_store = worker.store
+        # TODO: There should be a better way to get the other message store
+        # interfaces.
+        riak = self.query_store.manager
```
Comment (JayH5) [inline vumi_message_store/api/tests/test_message_export_worker.py]: This seemed pretty dirty to me... thoughts?

Comment (jerith) [inline vumi_message_store/api/tests/test_message_export_worker.py]: We could just create our own using `self.persistence_helper`, I suppose.

File: vumi_message_store/api/tests/test_message_export_worker.py [hunk 5]
```diff
@@ -0,0 +1,494 @@
...
+class TestMessageExportWorker(VumiTestCase):
+
+    @inlineCallbacks
+    def setUp(self):
+        self.persistence_helper = self.add_helper(
+            PersistenceHelper(use_riak=True))
+        self.worker_helper = self.add_helper(WorkerHelper())
+        self.msg_helper = self.add_helper(MessageHelper())
+
+        riak, redis = yield self.create_managers()
```
Comment (JayH5) [inline vumi_message_store/api/tests/test_message_export_worker.py:29]: Not sure if actually better :confused:

File: vumi_message_store/api/tests/test_message_export_worker.py [hunk 6]
```diff
@@ -0,0 +1,494 @@
...
+        self.operational_store = OperationalMessageStore(riak, redis)
+        self.batch_manager = MessageStoreBatchManager(riak, redis)
+
+    @inlineCallbacks
+    def create_managers(self):
+        riak = yield self.persistence_helper.get_riak_manager()
+        redis = yield self.persistence_helper.get_redis_manager()
+        self.add_cleanup(self.close_managers, riak, redis)
+        returnValue((riak, redis))
```
Comment (jerith) [inline vumi_message_store/api/tests/test_message_export_worker.py:39]: Instead of a separate `close_managers` method that closes both at once, why not just `self.add_cleanup(thing.close_manager)` when you create each manager?

Comment (JayH5) [inline vumi_message_store/api/tests/test_message_export_worker.py:39]: Ah didn't know I could do that... more deferred confusion :disappointed:

**Comment by jerith** [comment]
:+1:
